### PR TITLE
Allow user to manually input password when bio-metric fails

### LINKED ISSUE #19488: In case of many too many failed biometric attempts user is unable to confirm via password [ANDROID]
Android only. 

Reproducible in the following flows: sending community request, creating wallet, revealing sync QR code. 

**Steps:** 

1. Enable Fingerprint in app settings
2. Open sending Request to join community screen (creating wallet, revealing sync QR code)
3. Scan wrong finger to generate 'Too many attempts error' 
4. See if you will be able to confirm via password

**Expected result:** password bottom sheet should be opened so user be able to confirm via password. 

**Actual result:** error appears every time user tries to send request, password bottom sheet does not appear. 

https://github.com/status-im/status-mobile/assets/97245802/9eed7af3-bab8-4921-941e-b435a78c1ff7

### Additional Information

- Status version: nightly 
- Operating System: Android 12, Samsung Galaxy A52

### PR DESCRIPTION
**Pull request by Parveshdhull**
fixes https://github.com/status-im/status-mobile/issues/19488

### Video

https://github.com/status-im/status-mobile/assets/17097240/4140323f-6725-439a-96ca-1309ba3b502a

status: ready

### CODE REVIEW

File: src/status_im/common/standard_authentication/events.cljs [hunk 1]
```diff
@@ -33,6 +33,8 @@
                                             args-with-biometric-btn])
              :on-success     #(rf/dispatch [:standard-auth/on-biometric-success on-auth-success])
              :on-fail        (fn [err]
+                               (rf/dispatch [:standard-auth/authorize-with-password
```
Comment (Parveshdhull) [inline src/status_im/common/standard_authentication/events.cljs:36]:
```
(when (and err
           (string/includes?
            (string/lower-case
             (:orig-error-message (.-data err)))
            "too many attempts"))
```
I initially added a condition here to only show password input when too many attempt. But I think user should be able to manually input password when bio-metric fails, even if due to some other reason.

**Comment by Parveshdhull** [comment]
@clauxx is this leftover, or by choice

https://github.com/status-im/status-mobile/blob/f50ca4606f0e92ffc45900e468139242aa44fb3e/src/react_native/biometrics.cljs#L44

image

I mean, we have access to original error "Too many failed attempts", So I think we should show that to user instead of unknown error. wdyt?

**Comment by J-Son89** [comment]
@Parveshdhull - would be nice to include a video/ images if this affects the UI in any way ? 🤔 
Would be nice to see the behaviour of the flow etc 👍

File: src/status_im/common/standard_authentication/events.cljs (hunk 1 above)
Comment (J-Son89) [reply src/status_im/common/standard_authentication/events.cljs:36]: yeah makes sense 👍

**Comment by Parveshdhull** [comment]
Thank you @J-Son89, updated the description with the video.

btw, video shows unknown error, but as I mentioned in https://github.com/status-im/status-mobile/pull/19989#issuecomment-2106255606, we have access to actual error. So maybe can show "Too many attempts" instead of unknown error. We can address that later.

**Comment by J-Son89** [comment]
thanks @Parveshdhull - yep it would be great to have the proper error as unknown just looks like something went wrong but not necessarily the users fault. Fine to do in a follow up 👍

**Review by J-Son89 — APPROVED** [review]

**Comment by clauxx** [comment]
> @clauxx is this leftover, or by choice
> 
> https://github.com/status-im/status-mobile/blob/f50ca4606f0e92ffc45900e468139242aa44fb3e/src/react_native/biometrics.cljs#L44
> 
> image
> 
> I mean, we have access to original error "Too many failed attempts", So I think we should show that to user instead of unknown error. wdyt?

Sorry for the late reply. No it wasn't intended, this solution makes sense, thanks :+1:

**Review by clauxx — APPROVED** [review]

**Comment by qoqobolo** [comment]
Hey @Parveshdhull, thanks for the fix!
Everything looks good to me, but could you clarify one thing regarding the error, please?


If a user fails all biometric attempts and then enters a password manually, e.g. when creating a wallet account, then the next time they need to use biometrics again (e.g. for creating another wallet account or even relogin), they won't be able to use it and will immediately get this error again.
This can only be fixed by entering the device password (i.e. by locking the device and entering the password or in settings by disabling/enabling touch ID).

So two questions:
1. Is this expected? (it seems so)
2. You discussed that we should show `Too many attempts` error instead of the current `unknown` one, but will that `Too many attempts` error also inform the user that they will need to enter the device password to make biometric work again inside the app? 'Cause now it doesn't seem obvious to me as a user. 

Note: on iOS I can use biometrics again after getting the error and entering my password manually 

Steps to reproduce to make it more clear:
1. Start creating a new wallet account
2. Fail biometric authentification
3. Enter password
4. Try to create another wallet account
5. Try to relogin to the app

https://github.com/status-im/status-mobile/assets/67952253/634304b1-84d2-439e-b166-047e736be9ce

Screenshot 2024-05-17 at 15 09 19

**Comment by Parveshdhull** [comment]
hi @qoqobolo, thank you very much for testing the PR.

> This can only be fixed by entering the device password (i.e. by locking the device and entering the password or in settings by disabling/enabling touch ID).

I think it also gets fixed by itself after some time :thinking: 

> You discussed that we should show Too many attempts error instead of the current unknown one, but will that Too many attempts error also inform the user that they will need to enter the device password to make biometric work again inside the app?

No, we will only show error (too many attempts), or probably probably better if we could show "Too many attempts, try after some time or enter password manually". wdyt?

Also about informing user process making it work by entering device password etc.. I think we probably should not give this information to user, because biometric is os/device specific. We can't be sure, what will work on which device.

**Comment by qoqobolo** [comment]
>  or probably probably better if we could show "Too many attempts, try after some time or enter password manually". wdyt?

This sounds good to me 👍

Thanks for the clarification and for updating the [issue](https://github.com/status-im/status-mobile/issues/20007)!
PR can be merged.

**Pull request merged**